Fault array, fix retry, new parameter in Join activity

smartwaystudio opened the pull request:
**This PR has breaking changes**

Hi, I will try to set up my needs to explain what is this PR about. We are migrating our app core from custom Hangfire jobs to Elsa workflows with custom activities to enrich our app and also to have the ability to modify the workflows without making changes on the code. So, the first thing I did was to copy my hangfire flow to custom activities on Elsa. Everything was OK and the workflow was being executing with the same results. Then, I started do dig more, so, I purpossely failed some activities (custom and fault activity) to try the retry functionality and it wasn't working. Already opened an [issue with the problem ](https://github.com/elsa-workflows/elsa-core/issues/3180). If I couldn't retry my failed activities, I would need to give up elsa, so I started to checking the core code. I saw that Workflowinstance.CurrentActivity was always null, so the reviver was always using the first workflow activity as retry. But also found another problem (at least in my scenario). Elsa was only saving to database the last fault activity, so, in a loop or fork, just the last fault activity was being tagged as faulted, the others were OK. I changed the fault attribute to faults as SimpleStack. Now, I can save all the faulted activities. And, then, changed the reviver to check if there are faulted activities to schedule on retry. Now it's working good, even the retry. Now you can retry all the faulted activities inside a loop. 
![imagen](https://user-images.githubusercontent.com/30892914/177147010-6e01ffbb-5dba-4ffc-be21-160c07a95aea.png)
![imagen](https://user-images.githubusercontent.com/30892914/177147065-c700075c-0598-4680-94db-55834f52e5f8.png)

(Be careful with how you design your loop. If the activities input are dependant on other activities, for example, the output of the loop, it won't work as expected, because it will use the last output of the loop as input for every activity that uses "return foreach.Output()". The way to solve this is to transmit the input from the first activity in the loop to the input of the next activity, just returning it as output.)

Also, I found another problem. In a fork scenario, the final join continues the execution (with both configuration modes) even if the inbound activities has failed (tagged as faulted). I added a new parameter (bool) to the Join activity that prevents that. It checks if there are fault or scheduled activities (this change is chained with the previous) and stop the execution. Just changed IsDone method. Now, it will add to the journal the activities that have faulted, like the inbounds journal data.
![imagen](https://user-images.githubusercontent.com/30892914/177146831-18f18d44-6e75-4e3a-9db2-18e7a7d02f33.png)

There are 2 activity list checked on isDone. Faults is used to know if there are some faults in the current workflowinstance execution. The Scheduled activity list is used on retry. When you retry a faulted workflow, it moves the activities from Faults to scheduled list.

Finally, I updated the designer with the new Faults parameter (was Fault) and now, all the faulted activities in the designer are tagged red.
![imagen](https://user-images.githubusercontent.com/30892914/177143457-4e26e3a5-1cbf-4c50-9158-23dd5e6210de.png)

dnfadmin commented:
[![CLA assistant check](https://cla.dotnetfoundation.org/pull/badge/not_signed)](https://cla.dotnetfoundation.org/elsa-workflows/elsa-core?pullRequest=3182) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/elsa-workflows/elsa-core?pullRequest=3182) before we can accept your contribution.<br/><br/>:x: smartwaystudio [sign now](https://cla.dotnetfoundation.org/elsa-workflows/elsa-core?pullRequest=3182)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/elsa-workflows/elsa-core?pullRequest=3182) it.</sub>

smartwaystudio commented:
Willing to add more changes to this PR asap about the "Run workflow" activity because I just found that If the secondary workflow is faulted, the Run workflow activity is faulted, but if you try to retry the parent workflow, it generates a new secondary workflow instance instead of retrying the faulted one.

mohdali commented:
This looks very interesting and we definitely need more options for fault handling.

@sfmskywalker can you please take a look and advise as it has breaking changes.

smartwaystudio commented:
Hi, as I told, new changes have been pushed. The RunWorkFlow activity was always creating a new ChildWorkflow instance when retrying. There could be some scenario that you don't want that to happen, so I added a new Advanced parameter in the activity to, instead of creating new instances, retry the faulted when retrying.
![image](https://user-images.githubusercontent.com/30892914/177963531-10d4bfa2-1bdc-4330-b452-5bd0fcc9ac5a.png)

There is a general "problem" in all activities. The problem is that they are not unique in a branched workflow. Inside a loop, the same activity instance is used for all the iterations, so is hard to determine "who am I". That was happening also inside runworkflow. If not in a branched workflow, the retry was easy, but inside a loop, you can't determine what child workflow needs to be revived, because the activity just store the last iteration state inside, so ChildWorkflowInstanceID was pointing to the last. So, what I've done is to generate a hash for the values of the Input and store it. When retrying, using the input hash I can get the real ChildWorkflowInstance id and retry it.

![image](https://user-images.githubusercontent.com/30892914/177963385-e8aae4e6-538e-4c71-93fa-261201b32602.png)
![image](https://user-images.githubusercontent.com/30892914/177964340-008cf97a-64d4-48c1-8c3b-9a62e2fe9b60.png)


Here is a test without and with that parameter active. I just have a 4 iterations loop that fault. If I retry without that option, I get another 4 instances of the ChildWorkflow:
![image](https://user-images.githubusercontent.com/30892914/177963760-d3969b1b-3d9c-4f32-96ec-d4fa6d51f3c8.png)

If the option is active, it retry the faulted (I changed the Fault activity code to fast testing, returning done):
![image](https://user-images.githubusercontent.com/30892914/177964025-02a36cb6-fa36-4fea-b27e-723cd7beff3c.png)

sfmskywalker commented:
Great findings and solutions!

> There is a general "problem" in all activities. The problem is that they are not unique in a branched workflow. Inside a loop, the same activity instance is used for all the iterations, so is hard to determine "who am I"

Exactly. Your fix for this with the RunWorkflow activity is clever. The "real" fix might be for each activity execution to have its own instance ID - but that might require some significant refactorings, perhaps via https://github.com/elsa-workflows/elsa-core/issues/1059 - obviously out of scope for this PR

As an FYI: Elsa 3 already handles this today where each activity execution acquires its ow unique instance ID.

sfmskywalker commented on src/core/Elsa.Core/Activities/Workflows/RunWorkflow/RunWorkflow.cs:
```diff
@@ -98,22 +104,94 @@ public string ChildWorkflowInstanceId
+                }
+                childWorkflowStatus = childWorkflowInstance.WorkflowStatus;
+                ChildWorkflowInstanceId = childWorkflowInstance.Id;
+
+                
```
I noticed a few excessive blank lines here and there - let's tidy it up 😄

smartwaystudio replied:
My bad 😉 my brain was struggling hard with this Activity. Cleaning it next commit

sfmskywalker replied:
I can relate 😄

sfmskywalker commented on src/designer/elsa-workflows-studio/src/components/screens/workflow-instance-viewer/elsa-workflow-instance-journal/elsa-workflow-instance-journal.tsx:
```diff
@@ -242,7 +243,7 @@ export class ElsaWorkflowInstanceJournal {
                 <div>
                   <span
-                    class="elsa-h-8 elsa-w-8 elsa-rounded-full elsa-bg-green-500 elsa-flex elsa-items-center elsa-justify-center elsa-ring-8 elsa-ring-white elsa-mr-1"
+                    class={`elsa-h-8 elsa-w-8 elsa-rounded-full elsa-bg-${eventColor}-500 elsa-flex elsa-items-center elsa-justify-center elsa-ring-8 elsa-ring-white elsa-mr-1`}
```
I am guilty of this as well, but we shouldn't dynamically construct TailwindCSS classes like this, because they will be filtered out by the CSS cleanup phase in the build action.

Instead, it's better to assign the complete class value to the `eventColor` variable. Example (pseudo code):

`const eventColor = faulted ? 'elsa-bg-red-500' : 'elsa-bg-green-500'`

This way, the complete TailwindCSS class strings will be recognized.

smartwaystudio replied:
Nice, changing it.

smartwaystudio commented on src/core/Elsa.Core/Activities/ControlFlow/Join/Join.cs:
```diff
@@ -47,6 +47,10 @@ public enum JoinMode
 
+
+        [ActivityInput(Hint = "True if incoming activities executed should not be Faulted to continue.", UIHint = ActivityInputUIHints.SingleLine, SupportedSyntaxes = new[] { SyntaxNames.JavaScript, SyntaxNames.Liquid })]
+        public bool WaitFaulted { get; set; } = default!;
```
I agree that stopping when faulted should be the correct way to do the Join Activity. Changing it to be the default behaviour and removing the property in next commit.

sfmskywalker replied:
Perfect.

sfmskywalker commented on src/core/Elsa.Abstractions/Services/Models/WorkflowExecutionContext.cs:
```diff
@@ -188,7 +188,7 @@ public void Fault(Exception? exception, string message, string? activityId, obje
             var clock = ServiceProvider.GetRequiredService<IClock>();
             WorkflowInstance.WorkflowStatus = WorkflowStatus.Faulted;
             WorkflowInstance.FaultedAt = clock.GetCurrentInstant();
-            WorkflowInstance.Fault = new WorkflowFault(SimpleException.FromException(exception), message, activityId, activityInput, resuming);
```
> Maybe I can think something about moving the Fault value to the new Faults property so everything will be aligned

Exactly, that would be sweet.

Thanks man, appreciate your efforts!

smartwaystudio commented:
> Great findings and solutions!
> 
> > There is a general "problem" in all activities. The problem is that they are not unique in a branched workflow. Inside a loop, the same activity instance is used for all the iterations, so is hard to determine "who am I"
> 
> Exactly. Your fix for this with the RunWorkflow activity is clever. The "real" fix might be for each activity execution to have its own instance ID - but that might require some significant refactorings, perhaps via #1059 - obviously out of scope for this PR
> 
> As an FYI: Elsa 3 already handles this today where each activity execution acquires its ow unique instance ID.

When I started to design some testing workflows to learn about Elsa I was a bit surprised when the execution log in a loop was showing the same ActivityId for one looped activity and also just storing the last executed activity value. Moreover, that means that the parallelism is limited. I thought that, using hangfire, it would create a hangfire job for each branch, at least in the parallel foreach. That issue you are talking about would be very nice but indeed, requires a lot of refactoring. Can you point me how have you done this in Elsa 3? Maybe I can think about it! 😄

sfmskywalker reviewed (approved):
Awesome job 👍🏻

mohdali commented:
This is really exciting and I want to test it soon. @smartwaystudio is it possible to share the json of the test cases you created?

mohdali commented:
> > This is really exciting and I want to test it soon. @smartwaystudio is it possible to share the json of the test cases you created?
> 
> What I've done to test was creating some workflow (Fork, foreach, pforeach, while, fork-join, runWorkflow, foreach-runworkflow) and temporary modifying the Fault activity. The first run, the workflow fails and then, I change the Fault activity to return Done() (also adding a outcome to connect an activity to it) to test the retry. I've done this way because there is no way to edit the current version of workflow with dashboard.
> 
> Do you want a export of those workflow with the modified fault?

Yes please share the export. I'm not sure how you made the modification but I think if you unpublish then republish the workflow it will modify the current version.

smartwaystudio commented:
> > > This is really exciting and I want to test it soon. @smartwaystudio is it possible to share the json of the test cases you created?
> > 
> > 
> > What I've done to test was creating some workflow (Fork, foreach, pforeach, while, fork-join, runWorkflow, foreach-runworkflow) and temporary modifying the Fault activity. The first run, the workflow fails and then, I change the Fault activity to return Done() (also adding a outcome to connect an activity to it) to test the retry. I've done this way because there is no way to edit the current version of workflow with dashboard.
> > Do you want a export of those workflow with the modified fault?
> 
> Yes please share the export. I'm not sure how you made the modification but I think if you unpublish then republish the workflow it will modify the current version.

[test_workflows.zip](https://github.com/elsa-workflows/elsa-core/files/9095199/test_workflows.zip)

Here you have, some json workflows and `Fault.cs`. What I do is to force force Fault activity to sometimes `return Done()`. First execution of the workflow I return fault or exception and then, I change the code to `return Done()` to verify that the workflow ends properly. Is like emulating a scenario when an activity suddenly fails and you just need to retry it (ex: because it was a network error)

mohdali commented:
> 

@smartwaystudio thanks for the examples!

I have tested with the scenario of having dependent activities in the foreach and they executed once the workflow is retired and the error is cleared. This is really awesome!

![image](https://user-images.githubusercontent.com/4145465/179266239-030b24ce-dce3-4875-9cf3-3e7017eb84e9.png)
